collator/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+
+//! Collation Logic.
+//!
+//! A collator node lives on a distinct parachain and submits a proposal for
+//! a state transition, along with a proof for its validity
+//! (what we might call a witness or block data).
+//!
+//! One of collators' other roles is to route messages between chains.
+//! Each parachain produces a list of "egress" posts of messages for each other
+//! parachain on each block, for a total of N^2 lists all together.
+//!
+//! We will refer to the egress list at relay chain block X of parachain A with
+//! destination B as egress(X)[A -> B]
+//!
+//! On every block, each parachain will be intended to route messages from some
+//! subset of all the other parachains.
+//!
+//! Since the egress information is unique to every block, when routing from a
+//! parachain a collator must gather all egress posts from that parachain
+//! up to the last point in history that messages were successfully routed
+//! from that parachain, accounting for relay chain blocks where no candidate
+//! from the collator's parachain was produced.
+//!
+//! In the case that all parachains route to each other and a candidate for the
+//! collator's parachain was included in the last relay chain block, the collator
+//! only has to gather egress posts from other parachains one block back in relay
+//! chain history.
+//!
+//! This crate defines traits which provide context necessary for collation logic
+//! to be performed, as the collation logic itself.
+
+extern crate futures;
+extern crate polkadot_primitives as primitives;
+
+use std::collections::{BTreeSet, BTreeMap};
+
+use futures::{stream, Stream, Future, IntoFuture};
+use primitives::parachain::{self, ConsolidatedIngress, Message, Id as ParaId};
+
+/// Parachain context needed for collation.
+///
+/// This can be implemented through an externally attached service or a stub.
+pub trait ParachainContext {
+	/// Produce a candidate, given the latest ingress queue information.
+	fn produce_candidate<I: IntoIterator<Item=(ParaId, Message)>>(
+		&self,
+		ingress: I,
+	) -> parachain::RawProof;
+}
+
+/// Relay chain context needed to collate.
+/// This encapsulates a network and local database which may store
+/// some of the input.
+pub trait RelayChainContext {
+	type Error;
+
+	/// Future that resolves to the un-routed egress queues of a parachain.
+	/// The first item is the oldest.
+	type FutureEgress: IntoFuture<Item=Vec<Vec<Message>>, Error=Self::Error>;
+
+	/// Provide a set of all parachains meant to be routed to at a block.
+	fn routing_parachains(&self) -> BTreeSet<ParaId>;
+
+	/// Get un-routed egress queues from a parachain to the local parachain.
+	fn unrouted_egress(&self, id: ParaId) -> Self::FutureEgress;
+}
+
+/// Collate the necessary ingress queue using the given context.
+// TODO: impl trait
+pub fn collate_ingress<'a, R>(relay_context: R)
+	-> Box<Future<Item=ConsolidatedIngress, Error=R::Error> + 'a>
+	where R: RelayChainContext,
```

Comment:
Style nitpick: I find this formatting style of where clauses particularly unappealing :)
- We use tabs for identation, so this has this weird `--..` indentation
- You cannot easily re-order or copy bounds, since the first one contains `where` and seems like it's more important
- The last one is missing a comma, which makes it special as well.